JotunnLib/Managers/DungeonManager.cs

```diff
@@ -0,0 +1,387 @@
+
+            if (Rooms.Count > 0)
+            {
+                hashToName.Clear();
+
+                List<string> toDelete = new List<string>();
+                Logger.LogInfo(string.Format("Registering {0} custom rooms", Rooms.Count));
+                foreach (CustomRoom customRoom in Rooms.Values)
+                {
+                    try
+                    {
+                        Logger.LogDebug(string.Format("Adding custom room {0} with {1} theme", customRoom.Name, customRoom.Theme == 0 ? customRoom.ThemeName : customRoom.Theme.ToString()));
+                        if (customRoom.FixReference)
+                        {
+                            customRoom.Prefab.FixReferences(true);
+                            customRoom.FixReference = false;
+                        }
+
+                        if (customRoom.Theme != 0)
+                        {
+                            RegisterRoomInDungeonDB(customRoom);
+                        }
+                    }
+                    catch (MockResolveException ex)
+                    {
+                        Logger.LogWarning(string.Format("Skipping Room {0}: could not resolve mock {1} {2}", customRoom, ex.MockType.Name, ex.FailedMockName));
+                        toDelete.Add(customRoom.Name);
+                    }
+                    catch (Exception ex2)
+                    {
+                        Logger.LogWarning(string.Format("Exception caught while adding Room: {0}", ex2));
+                        toDelete.Add(customRoom.Name);
+                    }
+                }
+                foreach (string name in toDelete)
+                {
+                    Rooms.Remove(name);
+                }
+
+                DungeonDB.instance.GenerateHashList();
+                GenerateHashes();
+            }
+        }
+
+        private void OnDungeonGeneratorSetupAvailableRooms(DungeonGenerator self)
+        {
+            DungeonGeneratorTheme proxy = self?.gameObject.GetComponent<DungeonGeneratorTheme>();
+
+            if (DungeonGenerator.m_availableRooms != null)
+            {
+                if (proxy != null)
+                {
+                    Logger.LogDebug($"This dungeon generator has a custom theme = {proxy.m_themeName}, adding available rooms");
+
+                    DungeonGenerator.m_availableRooms.AddRange(Rooms.Values
+                        .Where(r => r.Room.m_enabled)
+                        .Where(r => r.ThemeName == proxy.m_themeName)
+                        .Select(r => r.RoomData));
+                }
+                else
+                {
+                    Logger.LogDebug($"Adding additional rooms of type {self.m_themes} to available rooms");
+
+                    DungeonGenerator.m_availableRooms.AddRange(Rooms.Values
+                        .Where(r => r.Room.m_enabled)
+                        .Where(r => self.m_themes.HasFlag(r.Theme))
+                        .Select(r => r.RoomData));
+                }
```

Comment:
Doesn't this branch potentially adds rooms twice to the `m_availableRooms` list, because it's already populated by the vanilla call?

Reply:
See explanation in this [comment](https://github.com/Valheim-Modding/Jotunn/pull/430#discussion_r1597048794) about how this works.